packages/SwingSet/src/kernel/kernel.js

```diff
@@ -832,30 +831,22 @@ export default function buildKernel(
       upgradeMessage,
       incarnationNumber: vatKeeper.getIncarnationNumber(),
     };
-    const disconnectObjectCapData = {
-      body: JSON.stringify(disconnectObject),
-      slots: [],
-    };
     /** @type { import('../types-external.js').KernelDeliveryStopVat } */
-    const kd1 = harden(['stopVat', disconnectObjectCapData]);
+    const kd1 = harden(['stopVat', kser(disconnectObject)]);
     const vd1 = vatWarehouse.kernelDeliveryToVatDelivery(vatID, kd1);
     const status1 = await deliverAndLogToVat(vatID, kd1, vd1);
 
     // make arguments for vat-vat-admin.js vatUpgradeCallback()
+    /**
+     * @param {SwingSetCapData} _errorCD
+     * @returns {RawMethargs}
+     */
     function makeFailure(_errorCD) {
-      insistCapData(_errorCD); // capdata(Error)
-      // const args = [upgradeID, false, JSON.parse(errorCD.body)];
+      insistCapData(_errorCD); // kser(Error)
+      // const args = [upgradeID, false, kunser(errorCD)];
       // actually we shouldn't reveal the details, so instead we do:
-      const error = {
-        '@qclass': 'error',
-        name: 'Error',
-        message: 'vat-upgrade failure',
-      };
-      const args = [upgradeID, false, error];
-      return {
-        body: JSON.stringify(['vatUpgradeCallback', args]),
-        slots: [],
-      };
+      const error = kser(Error('vat-upgrade failure'));
```

Comment:
oh, this might explain why you have the double-unserialize in some of the tests, in which case I think both sides need to remove a serialization layer

Reply:
I'm not sure.  There are places where double serialization is correct, when a capdata is part of some other message's payload.

Reply:
I've reviewed the various calls to `makeError` and verified that the output is used in a place that's expecting capdata.  I also added a test to check that the output from `makeError` and the output from `kser(Error(stuff))` is what things expect it to be; the later test will fail if the encoding changes and thereby flag when things need to be reinspected.